symbolic-base/src/ZkFold/Base/Data/Logical.hs

```diff
@@ -0,0 +1,220 @@
+  , GEq (..)
+  ) where
+
+import           Control.Category
+import qualified Data.Bool                        as H
+import           Data.Foldable                    hiding (all, and, any, elem, or)
+import           Data.Functor.Identity
+import qualified Data.Functor.Rep                 as R
+import           Data.Kind
+import qualified GHC.Generics                     as G
+import           Prelude                          (type (~))
+import qualified Prelude                          as H
+
+import           ZkFold.Base.Algebra.Basic.Number
+import           ZkFold.Base.Data.Vector
+
+class Boolean b where
+
+  true, false :: b
+
+  not :: b -> b
+
+  (&&), (||), xor :: b -> b -> b
+
+instance Boolean H.Bool where
+  true = H.True
+  false = H.False
+  not = H.not
+  (&&) = (H.&&)
+  (||) = (H.||)
+  xor = (H./=)
+
+all :: (Boolean b, Foldable t) => (x -> b) -> t x -> b
+all f = foldr ((&&) . f) true
+
+any :: (Boolean b, Foldable t) => (x -> b) -> t x -> b
+any f = foldr ((||) . f) false
+
+and :: (Boolean b, Foldable t) => t b -> b
+and = all id
+
+or :: (Boolean b, Foldable t) => t b -> b
+or = any id
+
+class Boolean (BooleanOf a) => Conditional a where
+
+  type BooleanOf a :: Type
+  type BooleanOf a = GBooleanOf (G.Rep a)
+
+  bool :: a -> a -> BooleanOf a -> a
+  default bool ::
+    ( G.Generic a
+    , GConditional (G.Rep a)
+    , BooleanOf a ~ GBooleanOf (G.Rep a)
+    ) => a -> a -> BooleanOf a -> a
+  bool f t b = G.to (gbool (G.from f) (G.from t) b)
+
+instance Conditional y => Conditional (x -> y) where
+  type BooleanOf (x -> y) = BooleanOf y
+  bool f t b x = bool (f x) (t x) b
+
+instance
+  ( Conditional x0
+  , Conditional x1
+  , BooleanOf x0 ~ BooleanOf x1
+  ) => Conditional (x0,x1)
+
+instance
+  ( Conditional x0
+  , Conditional x1
+  , Conditional x2
+  , BooleanOf x0 ~ BooleanOf x1
+  , BooleanOf x1 ~ BooleanOf x2
+  ) => Conditional (x0,x1,x2)
+
+instance
+  ( Conditional x0
+  , Conditional x1
+  , Conditional x2
+  , Conditional x3
+  , BooleanOf x0 ~ BooleanOf x1
+  , BooleanOf x1 ~ BooleanOf x2
+  , BooleanOf x2 ~ BooleanOf x3
+  ) => Conditional (x0,x1,x2,x3)
+
+instance (R.Representable v, Conditional x)
+  => Conditional (G.Generically1 v x) where
+    type BooleanOf (G.Generically1 v x) = BooleanOf x
+    bool (G.Generically1 vf) (G.Generically1 vt) b =
+      G.Generically1 (R.mzipWithRep (\f t -> bool f t b) vf vt)
+
+deriving via G.Generically1 (Vector n) x
+  instance (KnownNat n, Conditional x) => Conditional (Vector n x)
+
+instance Conditional (Identity x) where
+  type BooleanOf (Identity x) = H.Bool
+  bool = H.bool
+
+ifThenElse :: Conditional a => BooleanOf a -> a -> a -> a
+ifThenElse b t f = bool f t b
+
+class Conditional a => Eq a where
+
+  (==) :: a -> a -> BooleanOf a
+  default (==) ::
+    ( G.Generic a
+    , GEq (G.Rep a)
+    , BooleanOf a ~ GBooleanOf (G.Rep a)
+    ) => a -> a -> BooleanOf a
+  x == y = geq (G.from x) (G.from y)
+
+  (/=) :: a -> a -> BooleanOf a
```

Comment:
I wonder if `(/=)` can be defined as `x /= y = not (x == y)` as, thanks to recent optimizations by @hovanja2011, this should have the same number of constraints in circuits